packages/braintree-integration/src/braintree-fastlane/braintree-fastlane-utils.ts

```diff
@@ -0,0 +1,458 @@
+
+            await this.paymentIntegrationService.updatePaymentProviderCustomer({
+                authenticationState,
+                addresses,
+                instruments,
+            });
+
+            if (billingAddresses.length > 0) {
+                await this.paymentIntegrationService.updateBillingAddress(billingAddresses[0]);
+            }
+
+            if (shippingAddresses.length > 0 && cart.lineItems.physicalItems.length > 0) {
+                await this.paymentIntegrationService.updateShippingAddress(shippingAddresses[0]);
+            }
+        } catch (error) {
+            // TODO: we should figure out what to do here
+            // TODO: because we should not to stop the flow if the error occurs on paypal side
+        }
+    }
+
+    async runPayPalFastlaneAuthenticationFlowOrThrow(email?: string): Promise<void> {
+        try {
+            const methodId = this.getMethodIdOrThrow();
+
+            const braintreeFastlane = this.getBraintreeFastlaneOrThrow();
+            const { lookupCustomerByEmail, triggerAuthenticationFlow } = braintreeFastlane.identity;
+
+            const state = this.paymentIntegrationService.getState();
+            const cart = state.getCartOrThrow();
+            const customer = state.getCustomer();
+            const billingAddress = state.getBillingAddress();
+
+            const customerEmail = email || customer?.email || billingAddress?.email || '';
+
+            const { customerContextId } = await lookupCustomerByEmail(customerEmail);
+
+            if (!customerContextId) {
+                // Info: we should clean up previous experience with default data and related authenticationState
+                await this.paymentIntegrationService.updatePaymentProviderCustomer({
+                    authenticationState: BraintreeFastlaneAuthenticationState.UNRECOGNIZED,
+                    addresses: [],
+                    instruments: [],
+                });
+
+                this.browserStorage.setItem('sessionId', cart.id);
+
+                return;
+            }
+
+            const { authenticationState, profileData } = await triggerAuthenticationFlow(
+                customerContextId,
+            );
+
+            if (authenticationState === BraintreeFastlaneAuthenticationState.CANCELED) {
+                await this.paymentIntegrationService.updatePaymentProviderCustomer({
+                    authenticationState,
+                    addresses: [],
+                    instruments: [],
+                });
+
+                this.browserStorage.removeItem('sessionId');
+
+                return;
+            }
+
+            const shippingAddresses =
+                this.mapPayPalToBcAddress([profileData.shippingAddress]) || [];
+            const paypalBillingAddresses = this.getPayPalFastlaneBillingAddress(profileData);
+            const billingAddresses =
+                this.mapPayPalToBcAddress([paypalBillingAddresses as BraintreeFastlaneAddress]) ||
```

Comment:
We should avoid using `as` in our code if it's possible. I don't really understand why do you use it here, since `getPayPalFastlaneBillingAddress` returns `BraintreeFastlaneAddress | undefined`.

Reply:
I know, but ts-lint still gives an error